---
editor:
  markdown:
    wrap: 80
---

## Introduction

-   GIAB benchmark set background and history [@zook2019].
    -   Variant benchmarking and benchmark sets
    -   previous benchmark sets
        -   mapping-based
        -   assembly-based
        -   methods integration
    -   Limitations of previous versions [@zook2019; @krusche2019]
        -   mapping accessible
        -   assembly accuracy
        -   difficult regions (large tandem repeats, segmental duplications,
            ect.)
-   New high-accuracy de-novo assemblies enabling assembly-based variant
    benchmarks
    -   use of assemblies in previous benchmark sets
        -   v4.2.1, v0.6, CMRG, Tandem repeat [@english2025], XY benchmark
    -   Recent advances in sequencing and assembly methods
    -   Q100 HG002 assembly: background, accuracy, community effort and curation
-   Development of defrabb
    -   snakemake based
    -   assembly-based variant calling
    -   configurable VCF processing and annotations, benchmark region exclusions
    -   easily generate new benchmark sets for multiple reference genomes
-   Project overview
    -   Summary characterization of draft benchmark sets
    -   comparison to previous benchmark sets.
    -   External evaluation results

**NOTE: need to introduce variant v. genome benchmarking**