---
title: "v5q Benchmark Set Variant and Region Characterization"
date: '`r Sys.Date()`'
format:
  html:
    theme: flatly
    page-layout: full
    toc: true
    self-contained: true
  docx:default
editor: source
code-fold: true
execute:
  warning: false
---

## TODO

Analyses
- Look for small intervals in v5q



```{r setup}
#| include=FALSE,
#| echo=TRUE
library(tidyverse)
library(DT)
library(sessioninfo)
library(here)
library(assertthat)
library(patchwork)
library(gt)
# source(here("scripts", "happy_giab.R"))
```

## Utils/ Functions 
```{r utils}
get_var_allele_start <- function(rtg_stats_file) {
  stat_file_lines <- read_lines(rtg_stats_file)
  grep("Variant Allele Lengths", stat_file_lines)
}

get_rtg_metrics <- function(rtg_stats_file) {
  n_stat_rows <- -1
  var_len_start <- get_var_allele_start(rtg_stats_file)
  if (length(var_len_start) > 0) {
    n_stat_rows <- var_len_start - 3
  }

  read_delim(
    rtg_stats_file,
    n_max = n_stat_rows,
    skip = 1,
    delim = ":",
    trim_ws = TRUE,
    col_names = c("metric", "value"),
    col_types = "cc"
  )
}
```

## Variant Characterization

- Number of SNVs, INDELs, and SVs in v5 compared to v4 and v0.6
- Size distributions of INDELs and SVs in v5 compared to v4 and v0.6
- %TODO% Breakdown of variants in different genomic contexts (e.g., repeats, segmental duplications, etc.)
  - Stratifications v3.6
    - TR: LowComplexity/GRCh38_AllTandemRepeats.bed.gz
    - TR10kb: LowComplexity/GRCh38_AllTandemRepeats_ge101bp_slop5.bed.gz
    - HP: LowComplexity/GRCh38_AllHomopolymers_ge7bp_imperfectge11bp_slop5.bed.gz
    - SD: SegmentalDuplications/GRCh38_segdups.bed.gz
    - SD10kb: SegmentalDuplications/GRCh38_segdups_gt10kb.bed.gz
    - MAP: Mappability/GRCh38_lowmappabilityall.bed.gz


```{r}
#| warning: false
bench_var_rtg_stat_files <- list.files(
  path = here("results/vcfstats"),
  pattern = "vcfstats.txt",
  recursive = TRUE,
  full.names = TRUE
) %>%
  set_names(str_extract(., "(?<=vcfstats/).*(?=_vcfstats)"))

bench_var_metrics_df <- bench_var_rtg_stat_files %>%
  map_dfr(get_rtg_metrics, .id = "benchmarkset") %>%
  separate(
    col = benchmarkset,
    into = c("bench_version", "ref", "bench_type", "chrom"),
    sep = "_"
  ) %>%
  mutate(
    bench_version = case_when(
      bench_version == "v5q" ~ "v5.0q",
      bench_version == "v421" ~ "v4.2.1",
      bench_version == "v06" ~ "v0.6",
      bench_version == "cmrg" ~ "CMRG",
      bench_version == "tr" ~ "TR",
      TRUE ~ "FIXME"
    ),
    ref = case_when(
      ref == "grch37" ~ "GRCh37",
      ref == "grch38" ~ "GRCh38",
      ref == "chm13" ~ "CHM13v2.0",
      TRUE ~ "FIXME"
    ),
    bench_type = if_else(bench_type == "sv", "stvar", bench_type),
    bench_type = if_else(bench_version == "TR", "trvar", bench_type)
  )
```

Validate bench_var_metrics_df structure
```{r}
assert_that(
  is_tibble(bench_var_metrics_df),
  has_name(
    bench_var_metrics_df,
    c("bench_version", "ref", "bench_type", "metric", "value")
  ),
  nrow(bench_var_metrics_df) > 0,
  all(
    bench_var_metrics_df$bench_version %in%
      c("v5.0q", "v4.2.1", "v0.6", "CMRG", "TR")
  ),
  all(bench_var_metrics_df$ref %in% c("GRCh37", "GRCh38", "CHM13v2.0")),
  all(bench_var_metrics_df$bench_type %in% c("smvar", "stvar", "trvar"))
)
```

```{r}
diff_var_count_df <- list.files(path = here("results", "context_counts"),
                                 pattern = "_counts.tsv",
                                 recursive = TRUE,
                                 full.names = TRUE) %>%
  map_dfr(read_tsv, col_types = "cccci") %>%
  separate(benchmark, into = c("bench_version", "ref", "bench_type"), sep = "_") %>%
  mutate(
    bench_version = case_when(
      bench_version == "v5q" ~ "v5.0q",
      bench_version == "v421" ~ "v4.2.1",
      bench_version == "v06" ~ "v0.6",
      TRUE ~ "FIXME"
    ),
    ref = case_when(
      ref == "grch37" ~ "GRCh37",
      ref == "grch38" ~ "GRCh38",
      ref == "chm13" ~ "CHM13v2.0",
      TRUE ~ "FIXME"
    )
  )
```

```{r}
glimpse(diff_var_count_df)
```

### Variant Counts

Variant type definitions from rtg-tools documentation

- SNPs: can include variants with length greater than 1 bp but with only a mismatch in the first or last base.  
- MNPs: Cases where multiple bases change, but the lengths of the two alleles do not are considered to be MNPs
- Insertions/Deletions: Cases where there is pure addition or removal of bases are classified as Insertions or Deletions respectively
- Indels: there there is a length change between the REF and ALT, but it is not pure


```{r}
cnt_metrics <- c("SNPs", "MNPs", "Insertions", "Deletions", "Indels")
var_cnt_df <- bench_var_metrics_df %>%
  filter(metric %in% cnt_metrics) %>%
  mutate(value = as.integer(value))
```

Validate var_cnt_df structure
```{r}
assert_that(
  is_tibble(var_cnt_df),
  has_name(var_cnt_df, c("bench_version", "ref", "bench_type", "metric", "value")),
  nrow(var_cnt_df) > 0,
  is.integer(var_cnt_df$value),
  all(var_cnt_df$metric %in% cnt_metrics)
)
```
#### Small Variant Counts


```{r}
smvar_cnt_table <- var_cnt_df %>%
  ## MNPs filtered - all zeros
  filter(bench_type == "smvar", metric != "MNPs") %>%
  select(-bench_type) %>%
  ## %TODO% How to include insertions, deletions, INDELs
  mutate(metric = if_else(metric == "SNPs", "SNVs", "INDELs")) %>%
  pivot_wider(id_cols = c(ref, bench_version), names_from = metric, values_from = value, values_fn = sum) %>%
  mutate(Total = SNVs + INDELs)

smvar_cnt_table %>%
  arrange(ref) %>% gt::gt() %>% gt::fmt_integer(
    columns = c(SNVs, INDELs, Total)
  ) %>% 
  gt::cols_align(columns = c("ref", "bench_version"), 
                       align = "right") %>% 
  gt::cols_label(
    ref = "Reference",
    bench_version = "Benchmark",
    SNVs = "SNVs",
    INDELs = "INDELs",
    Total = "Total Variants"
  ) 

# gt::as_word(smvar_cnt_table %>%
#   arrange(ref) %>% gt::gt() %>% gt::fmt_integer(
#     columns = c(SNVs, INDELs, Total)
#   ) %>%
#   gt::cols_align(
#     columns = c("ref", "bench_version"),
#     align = "right"
#   ) %>%
#   gt::cols_label(
#     ref = "Reference",
#     bench_version = "Benchmark",
#     SNVs = "SNVs",
#     INDELs = "INDELs",
#     Total = "Total Variants"
#   ), 
#   here("results", "variant_counts", "smvar_counts_table.docx")
# )
```

#### Structural Variant Counts

```{r}
stvar_cnt_table <- var_cnt_df %>%
  filter(
    bench_type == "stvar",
    metric %in% c("Insertions", "Deletions", "Indels")
  ) %>%
  select(-bench_type) %>%
  pivot_wider(id_cols = c("ref", "bench_version"), names_from = metric, values_from = value, values_fn = sum) %>%
  mutate(Total = Insertions + Deletions + Indels)

stvar_cnt_table %>%
    arrange(ref) %>% gt::gt() %>% gt::fmt_integer(
    columns = c(Insertions, Deletions, Indels, Total)
  ) %>% 
  gt::cols_align(columns = c("ref", "bench_version"), 
                       align = "right") %>% 
  gt::cols_label(
    ref = "Reference",
    bench_version = "Benchmark"
  ) 

```

### Variant Size Distributions

%%TODO%% Check with JZ - do we want to add v0.6 to plots

```{r}
## TSV generate using make_sv_len.sh script
sv_len_df <- list.files(
  path = here("results", "sv_len"), pattern = "svlen.tsv", recursive = TRUE,
  full.names = TRUE
) %>%
  set_names(str_extract(., "(?<=sv_len/).*(?=_svlen.tsv)")) %>%
  map_dfr(read_tsv, col_types = "ciicciciic", na = ".", .id = "benchmarkset") %>%
  separate(col = benchmarkset, into = c("bench_version", "ref"), sep = "_") %>%
  filter(SVTYPE != ".") %>%
  filter(SVLEN > 49) %>%
  rename(CHROM = `#CHROM`) %>%
  mutate(
    TRF = case_when(TRF == "1" ~ "TR", is.na(TRF) ~ "nonTR", TRUE ~ "FIXME"),
    TRFlen = TRFend - TRFstart
  )

sv_len_fig_df <- sv_len_df %>%
  filter(SVLEN > 49) %>%
  mutate(SVLEN = if_else(SVTYPE == "DEL", -SVLEN, SVLEN)) %>%
  mutate(
    RepCat = case_when(
      TRF == "nonTR" & is.na(RM_clsfam) ~ "Non-Repeat",
      TRF == "TR" ~ "Tandem Repeat",
      str_detect(RM_clsfam, "SINE/Alu") ~ "Alu",
      str_detect(RM_clsfam, "LINE/L1") ~ "LINE",
      !is.na(RM_clsfam) ~ "Other Repeat",
      TRUE ~ "FIXME"
    ),
    ref = case_when(
      ref == "grch37" ~ "GRCh37",
      ref == "grch38" ~ "GRCh38",
      ref == "chm13" ~ "CHM13v2.0",
      TRUE ~ "FIXME"
    )
  )
```


Validate sv_len_df structure
```{r}
assert_that(
  is_tibble(sv_len_df),
  has_name(sv_len_df, c("CHROM", "POS0", "SVTYPE", "SVLEN", "TRF", "TRFstart", "TRFend", "TRFlen")),
  nrow(sv_len_df) > 0,
  is.integer(sv_len_df$SVLEN),
  all(sv_len_df$SVTYPE %in% c("INS", "DEL")),
  all(sv_len_df$TRF != "FIXME")
)

assert_that(
  is_tibble(sv_len_fig_df),
  all(sv_len_fig_df$RepCat != "FIXME")
)
```

```{r}
#| warning: false
(sv_len_dist_plt <- ggplot(
  sv_len_fig_df,
  aes(x = SVLEN)
) +
  geom_histogram(
    fill = "grey80", color = "grey20",
    bins = 50
  ) +
  geom_histogram(
    data = sv_len_fig_df %>% filter(RepCat %in% c("Alu", "LINE")),
    aes(fill = RepCat),
    color = "grey20",
    bins = 50, show.legend = FALSE
  ) +
  geom_text(aes(x = 300, y = 3000, label = "Alu"), size = 5, color = "darkblue") +
  geom_text(aes(x = -300, y = 3000, label = "Alu"), size = 5, color = "darkblue") +
  geom_text(aes(x = 6000, y = 500, label = "LINE"), size = 5, color = "darkorange") +
  geom_text(aes(x = -6000, y = 500, label = "LINE"), size = 5, color = "darkorange") +
  scale_x_continuous(
    breaks = c(-10000, -5000, -1000, -500, -100, -50, 0, 50, 100, 500, 1000, 5000, 10000),
    trans = scales::pseudo_log_trans(sigma = 50, base = 10),
    labels = scales::comma
  ) +
  scale_y_continuous(labels = scales::comma) +
  scale_fill_manual(values = c("Alu" = "darkblue", "LINE" = "darkorange")) +
  labs(
    x = "SV Length (bp)",
    y = "Count",
    fill = ""
  ) +
  facet_wrap(~ref, ncol = 1) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = -45, hjust = 0),
    minor_grid = element_blank()
  )
)
```
```{r}
ggsave(
  filename = here("manuscript", "figs", "sv_length_distribution.png"),
  plot = sv_len_dist_plt,
  width = 8,
  height = 6,
  dpi = 300
)
```

### Variants in Difficult Genomic Contexts

```{r}
diff_var_plt_df <- diff_var_count_df %>%
  filter(bench_type != "trvar",
         (bench_type == "stvar" & ref == "GRCh37") |
           (bench_type == "smvar" & ref == "GRCh38"),
         ## Not sure what to do about extra variant type - may also want to use a consistent method for getting variant counts, currently using vcfstats for whole genome and bcftools variant type annotations for diffficult variants
         variant_type %in% c("SNP","INDEL","INS","DEL"))
```

```{r}
(diff_smvar_plt <- diff_var_plt_df %>% filter(bench_type == "smvar") %>% 
  ggplot() + 
  geom_bar(aes(x = variant_type, y = count, fill = bench_version), 
           color = "grey20",
                      stat = "identity", position = "dodge") +
  facet_wrap( ~ context, nrow = 1) + 
  scale_y_continuous(label = scales::label_comma()) + 
      scale_fill_manual(values = c("v5.0q" = "darkgreen", "v4.2.1" = "darkorange", "v0.6" = "purple")) +
  labs(x = "Variant Type", y = "Variant Count", fill = "Benchmark Version", title = "Number of small variants (< 50bp) in difficult genomic context.") +
  theme_minimal() + 
  theme(legend.position = "bottom")
)
```
```{r}
(diff_stvar_plt <- diff_var_plt_df %>% filter(bench_type == "stvar") %>% 
  ggplot() + 
  geom_bar(aes(x = variant_type, y = count, fill = bench_version), 
           color = "grey20",
                      stat = "identity", position = "dodge") +
  facet_wrap( ~ context, nrow = 1) + 
  scale_y_continuous(label = scales::label_comma()) + 
   scale_fill_manual(values = c("v5.0q" = "darkgreen", "v4.2.1" = "darkorange", "v0.6" = "purple")) +
  labs(x = "Variant Type", y = "Variant Count", fill = "Benchmark Version", title = "Number of SVs (>= 50bp) in difficult genomic context.") +
  theme_minimal() + 
  theme(legend.position = "bottom")
)
```
```{r}
( diff_var_plt <- diff_smvar_plt + diff_stvar_plt +
  plot_layout(ncol = 1, guides = "collect") & theme(legend.position = "bottom")
)
```
```{r}
ggsave(
  filename = here("manuscript", "figs", "diff_variant_counts.png"),
  plot = diff_var_plt,
  width = 8,
  height = 6,
  dpi = 300
)
```


## Benchmark region characterization

- Number of bases in benchmark regions in v5 compared to v4 and v0.6

```{r}
bench_region_list <- list.files(
  path = here("data/benchmarksets"),
  pattern = "\\.bed(\\.gz)?$",
  recursive = TRUE,
  full.names = TRUE
) %>%
  set_names(str_extract(., "(?<=benchmarksets/).*"))
v5q_region_list <- list.files(
  path = here("data/20250117_v0.020_HG002Q100v1.1/draft_benchmarksets"),
  pattern = "benchmark.bed$",
  recursive = TRUE,
  full.names = TRUE
) %>%
  set_names(str_extract(., "(?<=benchmarksets/).*(?=-excluded/)"))

pg_region_list <- list(
  "platinum_pedigree_smvar_GRCh38" = here("data/platinum-pedigree-data/NA12878_hq_v1.2.smallvar.bed.gz"),
  "platinum_pedigree_stvar_GRCh38" = here("data/platinum-pedigree-data/NA12878_hq_v1.2.svs.bed.gz")
)
```


```{r}
regions_df <- c(bench_region_list, v5q_region_list, pg_region_list) %>%
  map_dfr(read_tsv, col_names = c("chrom", "start", "end"), col_types = "cii", .id = "benchmarkset") %>%
  mutate(
    interval_size = end - start,
    bench_version = case_when(
      str_detect(benchmarkset, "T2TQ100v1.1") ~ "v5q",
      str_detect(benchmarkset, "v4.2.1") ~ "v4.2.1",
      str_detect(benchmarkset, "v0.6") ~ "v0.6",
      str_detect(benchmarkset, "CMRG") ~ "CMRG",
      str_detect(benchmarkset, "TandemRepeats") ~ "TR",
      str_detect(benchmarkset, "platinum_pedigree") ~ "PG",
      TRUE ~ "FIXME"
    ),
    var_type = case_when(
      str_detect(benchmarkset, "smvar") | str_detect(benchmarkset, "smallvar") ~ "smvar",
      str_detect(benchmarkset, "stvar") | str_detect(benchmarkset, "SV") | str_detect(benchmarkset, "\\.svs\\.") ~ "stvar",
      str_detect(benchmarkset, "v4.2.1") ~ "smvar",
      str_detect(benchmarkset, "TandemRepeats") | str_detect(benchmarkset, "\\.trs\\.") ~ "TR",
      TRUE ~ "FIXME"
    ),
    ref = case_when(
      str_detect(benchmarkset, "v0.6") | str_detect(benchmarkset, "GRCh37") ~ "GRCh37",
      str_detect(benchmarkset, "GRCh38") | str_detect(benchmarkset, "platinum_pedigree") ~ "GRCh38",
      str_detect(benchmarkset, "CHM13") ~ "CHM13",
      TRUE ~ "FIXME"
    ),
    chrom = if_else(str_detect(chrom, "chr"), chrom, str_c("chr", chrom))
  )

# Validate regions_df structure and no FIXME values
assert_that(
  is_tibble(regions_df),
  has_name(regions_df, c("benchmarkset", "chrom", "start", "end", "interval_size", "bench_version", "var_type", "ref")),
  nrow(regions_df) > 0,
  !any(regions_df$bench_version == "FIXME"),
  !any(regions_df$var_type == "FIXME"),
  !any(regions_df$ref == "FIXME"),
  all(regions_df$interval_size > 0)
)
```

### Total Included Base Comparisons

- HG002Q100v1.1:[hg002v1.1.mat_Y_EBV_MT.fasta.gz.fai](https://s3-us-west-2.amazonaws.com/human-pangenomics/T2T/HG002/assemblies/hg002v1.1.mat_Y_EBV_MT.fasta.gz.fai), md5: c84f852b1cd4a00b12e6439ae7a2dd87, 
  - using total size of mat assembly from index file
- GRCh37:[hs37d5.fa.gz](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/references/GRCh37/hs37d5.fa.gz), md5: d10eebe06c0dbbcb04253e3294d63efc. 
  - Using seqkit to generate table of sequence lengths excluding Ns, `seqkit grep -r -p "^([1-9]|1[0-9]|2[0-2]|[XY])$" hs37d5.fa.gz | seqkit fx2tab > hs37d5_size.tsv`. Sanity check from NCBI [assembly stats](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.25_GRCh37.p13/GCF_000001405.25_GRCh37.p13_assembly_stats.txt) `{r 2861327195 - 2435416 - 3675142}`
- GRCh38: [GRCh38_GIABv3_no_alt_analysis_set_maskedGRC_decoys_MAP2K3_KMT2C_KCNJ18.fasta.gz](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/giab/release/references/GRCh38/GRCh38_GIABv3_no_alt_analysis_set_maskedGRC_decoys_MAP2K3_KMT2C_KCNJ18.fasta.gz), md5: 939ce19062d1462c09b88c55faca4d76. 
  - Using seqkit command `seqkit grep -r -p "^chr([1-9]|1[0-9]|2[0-2]|[XY])$" GRCh38_GIABv3_no_alt_analysis_set_maskedGRC_decoys_MAP2K3_KMT2C_KCNJ18.fasta.gz| seqkit fx2tab -inlsH -j 4 -C N > GRCh38_size.tsv`. 
  - Sanity check from NCBI [assembly stats](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_assembly_stats.txt) `{r 2937639396 - 6378305 - 4300658}`
- CHM13v2.0: [GCF_009914755.1_T2T-CHM13v2.0_genomic.fna.gz](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/914/755/GCF_009914755.1_T2T-CHM13v2.0/GCF_009914755.1_T2T-CHM13v2.0_genomic.fna.gz), md5: 9e6bf6b586bc8954208d1cc1d5f2fc99. 
  - Using seqkit command `seqkit grep -r -p "chromosome ([1-9]|1[0-9]|2[0-2]|[XY])$" GCF_009914755.1_T2T-CHM13v2.0_genomic.fna.gz]| seqkit fx2tab -inlsH -j 4 -C N > CHM13_size.tsv`. 
  - Sanity check from NCBI [assembly report](https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/009/914/755/GCF_009914755.1_T2T-CHM13v2.0/GCF_009914755.1_T2T-CHM13v2.0_assembly_stats.txt) 3117275501.

```{r}
hg002q100v1.1_size <- read_tsv(
  here("data/hg002v1.1.mat_Y_EBV_MT.fasta.gz.fai"),
  col_names = c("chrom", "length", "offset", "line_bases", "line_bytes"),
  col_types = "ciiii"
) %>%
  filter(chrom %in% paste0("chr", c(1:22, "X", "Y"), "_MATERNAL")) %>%
  summarise(genome_size = sum(length)) %>%
  pull(genome_size)

get_nongap_genome_size <- function(size_tsv) {
  read_tsv(
    size_tsv,
    col_names = c("chrom", "length", "Ns", "seq.hash"),
    col_types = "ciic",
    skip = 1
  ) %>%
    summarise(genome_size = sum(length) - sum(Ns)) %>%
    pull(genome_size)
}

GRCh37_size <- get_nongap_genome_size(here("data/hs37d5_size.tsv"))
GRCh38_size <- get_nongap_genome_size(here("data/GRCh38_size.tsv"))
CHM13_size <- get_nongap_genome_size(here("data/CHM13v2.0_size.tsv"))

ref_size <- data.frame(
  ref = c("GRCh37", "GRCh38", "CHM13", "HG002Q100v1.1"),
  genome_size = c(GRCh37_size, GRCh38_size, CHM13_size, hg002q100v1.1_size)
)

regions_df %>%
  filter(bench_version %in% c("v5q", "v4.2.1", "v0.6")) %>%
  left_join(ref_size) %>%
  mutate(
    chrom_group = if_else(
      chrom %in% paste0("chr", c("X", "Y")),
      "sexchrom",
      "autosome"
    )
  ) %>%
  group_by(var_type, ref, genome_size, chrom_group, bench_version) %>%
  summarise(total_bases = sum(interval_size), .groups = "drop") %>%
  pivot_wider(
    names_from = chrom_group,
    values_from = total_bases,
    values_fill = 0
  ) %>%
  mutate(
    total_bp = autosome + sexchrom,
    pct_ref_cvg = round(100 * (total_bp / genome_size), 0),
    pct_q100_cvg = round(100 * (total_bp / hg002q100v1.1_size), 0),
    total_bp = format(total_bp, big.mark = ","),
    autosome = format(autosome, big.mark = ","),
    sexchrom = format(sexchrom, big.mark = ",")
  ) %>%
  arrange(var_type, bench_version, ref) %>% 
  gt::gt()
```

Relative changes between v5q and v4.2.1/v0.6 by chromosome
```{r fig.width = 8, fig.height = 4}
(coverage_change_plt <- regions_df %>%
  filter(
    bench_version %in% c("v5q", "v4.2.1", "v0.6"),
    (var_type == "smvar" & ref == "GRCh38") |
      (var_type == "stvar" & ref == "GRCh37"),
    chrom %in% paste0("chr", c(1:22))
  ) %>%
  select(-starts_with("X")) %>%
  group_by(var_type, chrom, bench_version) %>%
  summarise(total_size = sum(interval_size), .groups = "drop") %>%
  mutate(
    vers = if_else(bench_version == "v5q", "new", "old"),
    chrom = factor(chrom, levels = paste0("chr", 1:22))
  ) %>%
  select(-bench_version) %>%
  pivot_wider(names_from = vers, values_from = total_size) %>%
  mutate(
    pct_change = 100 * (new - old) / old,
    total_change = new - old
  ) %>%
  ggplot() +
  geom_hline(yintercept = 0, color = "grey50") +
  geom_point(aes(x = chrom, y = pct_change, fill = total_change/1000000, shape = var_type),
    size = 4, color = "black", position = "jitterdodge"
  ) +
  scale_y_continuous(labels = scales::percent_format(scale = 1), n.breaks = 10) +
  scale_fill_gradient2(
    low = "red",
    mid = "white",
    high = "blue",
    midpoint = 0,
    labels = scales::comma
  ) +
  scale_shape_manual(values = c(21, 24)) +
  labs(
    x = "Chromosome", y = "Coverage Change",
    fill = "Coverage change (Mb)",
    shape = "Variant Type"
  ) +
  theme_minimal() +
  theme(
    axis.text.x = element_text(angle = -45, hjust = 0),
    legend.position = "bottom"
  )
)
```

```{r}
(ggsave(
  filename = here("manuscript", "figs", "benchmark_region_size_change.png"),
  plot = coverage_change_plt,
  width = 8,
  height = 4,
  dpi = 300
))
```

#### Inclusion of Difficult Regions

##### Difficult Region Bases Included
```{r}
diff_included_tbl <- read_tsv(here("results/context_coverage/all_context_coverage.tsv")) %>%
    separate(
      col = "benchmark",
      into = c("bench_version", "ref_id", "bench_type"),
      extra = "drop",
      sep = "_"
    ) %>%
    mutate(
      included_pct = round(100 * (overlap_bp / context_bp), 2)
    )

diff_included_plt_df <- diff_included_tbl %>%
  filter(
    (bench_version %in% c("v5q", "v421") & bench_type == "smvar" & ref_id == "grch38") |
      (bench_version %in% c("v5q", "v06") & bench_type == "stvar" & ref_id == "grch37"),
    chrom_scope == "autosomes"
  )
  
 
```

```{r}
ggplot(diff_included_plt_df) +
  geom_point(aes(x = bench_type, 
                 y = overlap_bp, 
                 fill = context, 
                 shape = bench_version), 
             position = "jitterdodge") +
  scale_y_log10() +
  annotation_logticks(sides = "lr") +
  scale_shape_manual(values = c(21, 24, 22)) +
  labs(x = "Benchmark Type", y = "Included Bases (bp)", fill = "Genomic Context", shape = "Benchmark Version") +
  facet_wrap(~context, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom", legend.box = "vertical") +
  guides(fill = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(shape = 21)))
```

```{r}
ggplot(diff_included_plt_df, aes(x = bench_type, 
                 y = included_pct)) +
  geom_point(aes( 
                 fill = context, 
                 shape = bench_version), 
             position = position_jitterdodge(dodge.width = 0.5)) +
  scale_shape_manual(values = c(21, 24, 22)) +
  scale_color_manual(values = c("grey20", "darkred")) + 
  scale_y_continuous(labels = scales::percent_format(scale = 1)) + 
  labs(x = "Benchmark Type", y = "Included Bases (%)", 
       fill = "Genomic Context", 
       shape = "Benchmark Version") +
  facet_wrap(~context, nrow = 1) +
  theme_minimal() +
  theme(legend.position = "bottom", 
        legend.box = "vertical") +
  guides(fill = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(shape = 21)))
```
```{r}
diff_included_plt_df %>% 
  select(ref, bench_type, context, bench_version, chrom_scope, included_pct) %>% 
  filter(chrom_scope == "autosomes") %>% 
  # group_by(bench_type) %>%
  # pivot_wider(names_from = bench_version, values_from = included_pct) #%>%
  # mutate(pct_change = (v5q - v421)/v421) %>%
  arrange(bench_type, context, chrom_scope) %>%
  gt::gt()

```


#### Benchmark Interval Size Distributions

Platinum pedigree truthsets downloaded from aws using `aws s3 sync --no-sign-request s3://platinum-pedigree-data/truthset_v1.2/ . 
```{r}
(interval_size_dist_plt <- regions_df %>%
  filter(
    bench_version %in% c("v5q", "v4.2.1", "v0.6", "PG"),
    (var_type == "smvar" & ref == "GRCh38") |
      (var_type == "stvar" & ref %in% c("GRCh37", "GRCh38")),
    chrom %in% paste0("chr", c(1:22))
  ) %>%
  select(-starts_with("X")) %>%
  ggplot() +
  geom_density(
    aes(x = interval_size, fill = bench_version),
    position = "identity",
    alpha = 0.5,
    bins = 100,
    outline.type = "upper"
  ) +
  scale_x_log10(labels = scales::comma) +
  annotation_logticks(sides = "b") +
  scale_fill_brewer(palette = "Set1") +
  facet_wrap(~var_type, ncol = 1) +
  theme_minimal() +
  theme(legend.position = "bottom") +
  labs(x = "Benchmark Interval Size (bp)", y = "Density", fill = "Benchmark Version"))
```
```{r}
(interval_count_plt <- regions_df %>%
  filter(
    bench_version %in% c("v5q", "v4.2.1", "v0.6", "PG"),
    (var_type == "smvar" & ref == "GRCh38") |
      (var_type == "stvar" & ref %in% c("GRCh37", "GRCh38")),
    chrom %in% paste0("chr", c(1:22))
  ) %>% 
  ggplot() + geom_bar(aes(x = bench_version, fill = bench_version), 
                      color = "grey20", position = "dodge",   alpha = 0.5) +
  facet_wrap(~var_type, nrow = 2, scales = "free") +
  labs(x = "Benchmark Version", y = "Number of Intervals", fill = "Benchmark Version") +
  scale_y_continuous(labels = scales::comma) +
  theme_minimal() +
  theme(legend.position = "none"))

```
```{r}
(interval_plt <- interval_size_dist_plt + 
  interval_count_plt
 ) 

ggsave(here("manuscript/figs/benchmark_intervals.png"), interval_plt,
         width = 8, height = 6)
```

Number of regions by benchmark set.
```{r}
regions_df %>%
  filter(
    bench_version %in% c("v5q", "v4.2.1", "v0.6", "PG"),
    (var_type == "smvar" & ref == "GRCh38") |
      (var_type == "stvar" & ref %in% c("GRCh37", "GRCh38")),
    chrom %in% paste0("chr", c(1:22))
  ) %>%
  group_by(bench_version, var_type, ref) %>%
  count() %>%
  mutate(bench_version = factor(bench_version, levels = c("v4.2.1", "v0.6",  "v5q", "PG"))) %>%
  arrange(var_type, bench_version) %>% 
  ungroup() %>% 
  select(var_type, ref, bench_version, n) %>% 
  gt::gt() %>% gt::tab_header(title = "Number of intervals in Benchmark Regions") %>% 
  gt::fmt_integer(columns = n) %>% 
  gt::cols_align(columns = c("var_type", "ref", "bench_version"), align = "right") %>% 
  gt::cols_label(
    var_type = "Variant Type",
    ref = "Reference",
    bench_version = "Benchmark",
    n = "Number of Intervals"
  )
```

## Medically Relevant Gene Comparison

- number of variants in medically relevant genes in v5 compared to v4 and v0.6 + CMRG
- Medically relevant gene coverage
Use code from CMRG paper as starting point https://github.com/usnistgov/cmrg-benchmarkset-manuscript/tree/master

## Exclusions

- __TODO__ Bases removed due to each exclusion (upset plot?)
- __TODO__ Variants removed due to each exclusion (upset plot?)
- __TODO__ Variants and bases not in v5 but in previous versions

```{r}
exclusion_intersect_files <- list.files(
  path = here("data/20250117_v0.020_HG002Q100v1.1/draft_benchmarksets"),
  pattern = "exclusion_intersection_summary.csv",
  recursive = TRUE,
  full.names = TRUE
) %>%
  set_names(str_remove(., ".*-excluded/"))

exclusion_intersect_df <- exclusion_intersect_files %>%
  map_dfr(read_csv, show_col_types = FALSE, .id = "benchmarkset") %>%
  separate(
    col = benchmarkset,
    into = c("ref", "assembly", "bench_type", "varcall"),
    extra = "drop",
    sep = "_"
  ) %>%
  select(-assembly, -varcall) %>%
  rename(
    dip_overlap_bp = bp,
    dip_excluded_pct = pct_of_initial
  ) %>%
  mutate(
    start_end = case_when(
      str_detect(genomic_region, "start") ~ "start",
      str_detect(genomic_region, "end") ~ "end",
      TRUE ~ ""
    ),
    genomic_region = str_remove(genomic_region, "_(slop|sorted).*bed$"),
    exclusion = case_when(
      genomic_region == "HG002Q100-errors" ~ genomic_region,
      genomic_region == "initial" ~ "diploid-regions",
      str_detect(genomic_region, "HG002Q100-") ~ str_extract(genomic_region, "(?<=HG002Q100-).*"),
      str_detect(genomic_region, "_dipcall-z2k_") ~ str_extract(genomic_region, "(?<=_dipcall-z2k_).*"),
      TRUE ~ genomic_region
    )
  )
```

```{r}
exclusion_gt_tbl <- exclusion_intersect_df %>%
  mutate(exclusion = trimws(paste(exclusion, start_end), "both")) %>%
  select(ref, exclusion, bench_type, dip_overlap_bp, dip_excluded_pct) %>%
  filter(!(exclusion %in% c("benchmark_regions", "diploid-regions"))) %>%
  pivot_wider(names_from = ref, values_from = c(dip_overlap_bp, dip_excluded_pct)) %>%
  arrange(-dip_overlap_bp_CHM13v2.0)

make_exclusion_tbl <- function(exclusion_tbl, table_title) {
  gt::gt(exclusion_tbl) %>%
    gt::fmt_integer(
      columns = c(dip_overlap_bp_GRCh37, dip_overlap_bp_GRCh38, dip_overlap_bp_CHM13v2.0)
    ) %>%
    gt::fmt_number(
      columns = c(dip_excluded_pct_GRCh37, dip_excluded_pct_GRCh38, dip_excluded_pct_CHM13v2.0),
      decimals = 2
    ) %>%
    gt::tab_spanner(
      label = "GRCh37",
      columns = c(dip_overlap_bp_GRCh37, dip_excluded_pct_GRCh37)
    ) %>%
    gt::tab_spanner(
      label = "GRCh38",
      columns = c(dip_overlap_bp_GRCh38, dip_excluded_pct_GRCh38)
    ) %>%
    gt::tab_spanner(
      label = "CHM13v2.0",
      columns = c(dip_overlap_bp_CHM13v2.0, dip_excluded_pct_CHM13v2.0)
    ) %>%
    gt::cols_label(
      dip_overlap_bp_GRCh37 = "bp",
      dip_excluded_pct_GRCh37 = "%",
      dip_overlap_bp_GRCh38 = "bp",
      dip_excluded_pct_GRCh38 = "%",
      dip_overlap_bp_CHM13v2.0 = "bp",
      dip_excluded_pct_CHM13v2.0 = "%"
    ) %>%
    gt::tab_header(
      title = table_title
    ) %>%
    gt::cols_align(columns = c(exclusion, bench_type), align = "right")
}
```

```{r}
combined_exclusion_tbl <- exclusion_intersect_df %>%
  mutate(exclusion = trimws(paste(exclusion, start_end), "both")) %>%
  select(ref, exclusion, bench_type, dip_overlap_bp, dip_excluded_pct) %>%
  filter(!(exclusion %in% c("benchmark_regions", "diploid-regions"))) %>%
  pivot_wider(names_from = ref, values_from = c(dip_overlap_bp, dip_excluded_pct)) %>%
  arrange(-dip_overlap_bp_CHM13v2.0) %>% 
  mutate(bench_type = case_when(exclusion == "total_excluded" ~ bench_type, 
                                exclusion == "svs-and-simple-repeats" ~ bench_type,
                                exclusion == "self-discrep" ~ bench_type,
                                TRUE ~ "both")) %>%
  distinct()
```


%%TODO%% Add variants to the exclusion tables
```{r}
combined_exclusion_tbl %>%
  make_exclusion_tbl(table_title = "Benchmark Set Exclusions")
```


